src/core/server/workspaces/workspace_permission_check.ts

```diff
@@ -0,0 +1,273 @@
+    if (!newPermissions) {
+      newPermissions = {};
+    }
+    switch (permissonType) {
+      case PERMISSION_TYPE_READ:
+        newPermissions.read = addPermissionToPrincipals(newPermissions.read, users, groups);
+        break;
+      case PERMISSION_TYPE_WRITE:
+        newPermissions.write = addPermissionToPrincipals(newPermissions.write, users, groups);
+        break;
+      case PERMISSION_TYPE_MANAGEMENT:
+        newPermissions.management = addPermissionToPrincipals(
+          newPermissions.management,
+          users,
+          groups
+        );
+        break;
+      case PERMISSION_TYPE_LIBRARY_READ:
+        newPermissions.library_read = addPermissionToPrincipals(
+          newPermissions.library_read,
+          users,
+          groups
+        );
+        break;
+      case PERMISSION_TYPE_LIBRARY_WRITE:
+        newPermissions.library_write = addPermissionToPrincipals(
+          newPermissions.library_write,
+          users,
+          groups
+        );
+        break;
+    }
+    return newPermissions;
+  }
+
+  public removePermission(
+    permissonType: string,
+    permissions: Permissions,
+    users?: string[],
+    groups?: string[]
+  ) {
+    if (!isParamsValid(permissonType, users, groups) || !permissions) {
+      return permissions;
+    }
+
+    const newPermissions = cloneDeep(permissions);
+    switch (permissonType) {
+      case PERMISSION_TYPE_READ:
+        newPermissions.read = deletePermissionFromPrincipals(newPermissions.read, users, groups);
+        break;
+      case PERMISSION_TYPE_WRITE:
+        newPermissions.write = deletePermissionFromPrincipals(newPermissions.write, users, groups);
+        break;
+      case PERMISSION_TYPE_MANAGEMENT:
+        newPermissions.management = deletePermissionFromPrincipals(
+          newPermissions.management,
+          users,
+          groups
+        );
+        break;
+      case PERMISSION_TYPE_LIBRARY_READ:
+        newPermissions.library_read = deletePermissionFromPrincipals(
+          newPermissions.library_read,
+          users,
+          groups
+        );
+        break;
+      case PERMISSION_TYPE_LIBRARY_WRITE:
+        newPermissions.library_write = deletePermissionFromPrincipals(
+          newPermissions.library_write,
+          users,
+          groups
+        );
+        break;
+    }
+    return newPermissions;
+  }
+
+  public genereateGetPermittedSavedObjectsQueryDSL(
+    permissonType: string,
+    savedObjectType?: string | string[],
+    user?: string,
+    group?: string
+  ) {
+    if ((!user && !group) || !permissonType || !PERMISSION_TYPES.has(permissonType)) {
+      return {
+        query: {
+          match_none: {},
+        },
+      };
+    }
+
+    const bool: any = {
+      filter: [],
+    };
+    if (!!user) {
+      bool.filter.push({
+        term: {
+          ['permissions.' + permissonType + '.users']: user,
+        },
+      });
+    } else if (!!group) {
```

Comment:
Could we add both user and group filter? Seems we won't go here by `else if`.

Reply:
Can login user and backend role be used at the same time?

Reply:
I am not sure if we will have this situation. From the signature of this function, it seems we can pass user and group together.